Add ndarray.getValuesAsync() method which returns a promise that resolves when values are ready.

nsthorat opened the pull request:
This implementation will poll the GPU for the availability of the disjoint query object as if we were benchmarking. The cost of polling is negligible, so we can use a setTimeout(fn, 0) to check.

I also removed the exponential backoff in the benchmarks, we can just poll as fast as possible since it's pretty cheap.

When the extension isn't enabled, we just call getValues and resolve immediately.

This change also allows flags to depend on each other, and values get cached so we don't have to construct lots of canvases when evaluating flags.

<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pair-code/deeplearnjs/146)
<!-- Reviewable:end -->

dsmilkov commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> Just few tiny comments. Great work!

---

Reviewed 14 of 14 files at r1.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[src/environment.ts, line 27 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/146#-Kug71iWfl9glR5prNyT:-Kug71iXScPC8L7tGmds:b-pedi6v) ([raw file](https://github.com/pair-code/deeplearnjs/blob/3c8d4c41c118c19c5751cfc7d19a266efa4d72ec/src/environment.ts#L27)):*
> ```XML
> 
> export interface Features {
>   'WEBGL_DISJOINT_QUERY_TIMER_EXTENSION_ENABLED'?: boolean;
> ```

Add short one liner comment on top of each feature explaining what each feature means (e.g. want to understand the diff between the two query time extensions)

---

*[src/environment_test.ts, line 42 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/146#-Kug8fRYomxupqVJsi7Z:-Kug8fRYomxupqVJsi7_:b5wws46) ([raw file](https://github.com/pair-code/deeplearnjs/blob/3c8d4c41c118c19c5751cfc7d19a266efa4d72ec/src/environment_test.ts#L42)):*
> ```XML
>                 return {loseContext: () => {}};
>               }
>               console.log('returning null');
> ```

remove console.log

---

*[src/util.ts, line 226 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/146#-Kug9NUuKorRUqBIMrCA:-Kug9NUuKorRUqBIMrCB:b-zhjxfn) ([raw file](https://github.com/pair-code/deeplearnjs/blob/3c8d4c41c118c19c5751cfc7d19a266efa4d72ec/src/util.ts#L226)):*
> (counter: number) => 0

this looks strange, shouldn't it be `(counter: number) => number` ?

---

*[src/util.ts, line 227 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/146#-Kug9CF9dw2te81YW6ez:-Kug9CF9dw2te81YW6f-:b-pytmj6) ([raw file](https://github.com/pair-code/deeplearnjs/blob/3c8d4c41c118c19c5751cfc7d19a266efa4d72ec/src/util.ts#L227)):*
> ```XML
> export function repeatedTry(
>     checkFn: () => boolean, delayFn = (counter: number) => 0,
>     maxCounter?: number) {
> ```

re-add the return type of `repeatedTry()`

---

*[src/math/webgl/gpgpu_context.ts, line 266 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/146#-KugBDN5JLV392LZsNDm:-KugBDN5JLV392LZsNDn:b-s2bk0w) ([raw file](https://github.com/pair-code/deeplearnjs/blob/3c8d4c41c118c19c5751cfc7d19a266efa4d72ec/src/math/webgl/gpgpu_context.ts#L266)):*
> ```XML
>    * the command buffer has finished executing the query.
>    * @param queryFn The query function containing GL commands to execute.
>    * @return a promise that resolves with the ellapsed time.
> ```

s/ellapsed time/ellapsed time in ms/

---

*[src/math/webgl/gpgpu_context.ts, line 311 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/146#-KugAp07dWZa5ZNlNcrI:-KugAp07dWZa5ZNlNcrJ:b-vhsfeh) ([raw file](https://github.com/pair-code/deeplearnjs/blob/3c8d4c41c118c19c5751cfc7d19a266efa4d72ec/src/math/webgl/gpgpu_context.ts#L311)):*
> ```XML
>                 .getQueryParameter(query, (this.gl as any).QUERY_RESULT);
>         // Return milliseconds.
>         resolve(timeElapsedMicros / 1000000);
> ```

same here: 1ms = 1000 micro seconds. Why divide by 1 million?

---

*[src/math/webgl/gpgpu_context.ts, line 349 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/146#-KugAYReL_iPpdv-bVUb:-KugAYReL_iPpdv-bVUc:b3578rp) ([raw file](https://github.com/pair-code/deeplearnjs/blob/3c8d4c41c118c19c5751cfc7d19a266efa4d72ec/src/math/webgl/gpgpu_context.ts#L349)):*
> timeElapsedMicros / 1000000

1 ms = 1000 micro seconds. Divide by 1000 instead of 1 million ?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/146#-:-KugBgnub9Sa0YCOwvZ2:b-yi9klc)*
<!-- Sent from Reviewable.io -->

nsthorat commented:
Thanks! Fixed your comments. It's so nice to have a thorough reviewer :)

---

Review status: 9 of 14 files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[src/environment.ts, line 27 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/146#-Kug71iWfl9glR5prNyT:-KujE9myreY0LbJdT1aT:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/3c8d4c41c118c19c5751cfc7d19a266efa4d72ec/src/environment.ts#L27)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

Add short one liner comment on top of each feature explaining what each feature means (e.g. want to understand the diff between the two query time extensions)
</blockquote></details>

Done.

---

*[src/environment_test.ts, line 42 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/146#-Kug8fRYomxupqVJsi7Z:-KujEXUIGHw9oVnvbaiD:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/3c8d4c41c118c19c5751cfc7d19a266efa4d72ec/src/environment_test.ts#L42)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

remove console.log
</blockquote></details>

Done.

---

*[src/util.ts, line 226 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/146#-Kug9NUuKorRUqBIMrCA:-KujEYGPJQQJaXzph90S:bn01jus) ([raw file](https://github.com/pair-code/deeplearnjs/blob/3c8d4c41c118c19c5751cfc7d19a266efa4d72ec/src/util.ts#L226)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

> (counter: number) => 0

this looks strange, shouldn't it be `(counter: number) => number` ?
</blockquote></details>

Ah it's not a type here. This argument is a function that maps the step counter to the timeout (so you could do exponential backoff). Here, we always return 0 so we have a setTimeout(fn, 0) loop

---

*[src/util.ts, line 227 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/146#-Kug9CF9dw2te81YW6ez:-KujEl2MdzSgYAfYHzOo:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/3c8d4c41c118c19c5751cfc7d19a266efa4d72ec/src/util.ts#L227)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

re-add the return type of `repeatedTry()`
</blockquote></details>

Done.

---

*[src/math/webgl/gpgpu_context.ts, line 266 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/146#-KugBDN5JLV392LZsNDm:-KujEw1Gf7NDM9l76xji:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/3c8d4c41c118c19c5751cfc7d19a266efa4d72ec/src/math/webgl/gpgpu_context.ts#L266)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

s/ellapsed time/ellapsed time in ms/
</blockquote></details>

Done.

---

*[src/math/webgl/gpgpu_context.ts, line 311 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/146#-KugAp07dWZa5ZNlNcrI:-KujF0CrgHCwrCO4WuNl:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/3c8d4c41c118c19c5751cfc7d19a266efa4d72ec/src/math/webgl/gpgpu_context.ts#L311)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

same here: 1ms = 1000 micro seconds. Why divide by 1 million?
</blockquote></details>

Done.

---

*[src/math/webgl/gpgpu_context.ts, line 349 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/146#-KugAYReL_iPpdv-bVUb:-KujF166I_V1gWVyxuSj:brmvkfy) ([raw file](https://github.com/pair-code/deeplearnjs/blob/3c8d4c41c118c19c5751cfc7d19a266efa4d72ec/src/math/webgl/gpgpu_context.ts#L349)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

> timeElapsedMicros / 1000000

1 ms = 1000 micro seconds. Divide by 1000 instead of 1 million ?
</blockquote></details>

Renamed it, it's actually nanoseconds :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/146#-:-KujFBWjZanLUsxLIpsh:b-ic4nlk)*
<!-- Sent from Reviewable.io -->
